### ISSUE TITLE
`datasummary_correlation` improvements

### ISSUE THREAD

**Issue by vincentarelbundock** [vincentarelbundock/modelsummary#227]
In PR #226  @joachim-gassen recommended 2 new features for `datasummary_correlation`:

> Thank you! Added. Another thing (would be another PR): I am thinking about implementing bolding based on p-values and a note to disclose the range of observations underlying the correlations. The latter would be useful for data with missing values. This would only work for standard methods and not for methods provided by function. Interested?

These are interesting suggestions. I'm definitely open to considering, but would like to think through the API a bit before any code is written. For instance,

### Bolding significance

The analogue here is the `stars` argument in `modelsummary`. Where the analogy breaks is that `stars` adds a *character* (the asterisk), whereas bolding requires a change in the *format*, not the content of the table.

All formatting in the package is outsourced to the `factory` function, which takes a data.frame and then dispatches it based on the `output` argument to `factory_kableExtra`, `factory_gt`, `factory_flextable`, `factory_huxtable`.

One core design decision is that I want users to do most of their table customization using those external packages. This gives flexibility for users, but also simplifies things considerably for `modelsummary`. To wit, there is currently no mechanism for bolding in any of the `factory_*` functions. 

Do we need to onboard a fully-featured bolding system to do this, or is there an intermediate route to make things easier *and* general for users?

### Note about range

This sounds like a useful idea.

My main concern with this feature is that I really want to limit the number of arguments, and keep those argument names as consistent as possible across the various functions of the package.

We already have a `notes` argument, and I wonder how it could be leveraged to achieve this with minimal changes and pain, while making things very easy for users.

Just brainstorming here, but maybe we could allow users to supply a function to the `notes` argument, and that function could be evaluated with all the arguments from the parent. So for instance,

```r
range_note <- function(data, ...) {
  out <- sprintf(
    "%s [%s, %s]",
    colnames(data),
    sapply(data, min),
    sapply(data, max))
  paste(out, collapse = ", ")
}
```

Then just:

```r
datasummary_correlation(mtcars, notes = range_note)
```

(This won't work now; just pseudo-code to brainstorm.)

**Comment by joachim-gassen** [comment]
Excellent points. In the meantime, I also had a more careful look at some of the inner workings of the package. I agree that keeping the API clean is important and that relegating table formatting to user space makes a lot of sense. This would imply that bolding for significance would be a user space task. Also, adding and modifying the table notes to provide information about the number of observations can be realised in user space. 

However, to increase the usability of `datasummary_correlation()`, I think there are two related points that would need to be addressed within the package since they affect the character content of the returned table. Both relate to how variable names are handled in the table:

- Correlation tables often contain multiple variables. If variable names are not very short, this tends to break the formatting because of the top row containing the variable names. Replacing the variable names in the top row by letter indicators addresses this but requires these letters to also be referenced in the first column
- Like in `modelsummary()` and in `datasummary()`, it would be good if users could provide speaking variable names as replacements for the variable names contained in the data

The below is an implementation that I whipped up because I needed it for my own work.  Maybe one could consider implementing the bits on `alpha_cols`, `vars_rename` and maybe `delete_diag` as well as `delete_upper_tria` in `datasummary()`? The remaining  bits are formatting and could well be handled in user land.

Here is the code that I currently use, together with some example code. 

```
library(modelsummary)
library(kableExtra)

kable_corr_table <- function(
  data,
  fmt = 2,
  title = NULL,
  notes = NULL,
  method = "pearson",
  var_rename = NULL,
  cols_alpha = FALSE,
  delete_diag = FALSE,
  delete_upper_tria = {if (method == "pearspear") FALSE else TRUE},
  bold = FALSE,
  n_note = FALSE,
  bold_sprintf = "Correlations with two-sided p-values below %.2f are printed in bold.",
  n_note_sprintf = "The number of observations ranges from %d to %d."
) {
  cor_list <- function(x, method) {
    mat <- as.matrix(x)
    n <- ncol(mat)
    out_r <- out_n <- out_p <- matrix(NA, n, n)
    diag(out_r) <- 1
    diag(out_p) <- 0.00
    diag(out_n) <- colSums(is.finite(mat))

    for (i in 1:(n - 1)) {
      for (j in (i + 1):n) {
        out_n[i, j] <- out_n[j, i] <- sum(is.finite(mat[, i]) & is.finite(mat[, j]))
        if (out_n[i, j] > 2) {
          tmp <- stats::cor.test(mat[, i], mat[, j], method = method, exact = FALSE)
          out_r[i, j] <- out_r[j, i] <- tmp$estimate
          out_p[i, j] <- out_p[j, i] <- tmp$p.value
        }
      }
    }
    colnames(out_r) <- rownames(out_r) <- colnames(mat)
    colnames(out_p) <- rownames(out_p) <- colnames(mat)
    colnames(out_n) <- rownames(out_n) <- colnames(mat)
    list(r = out_r, p = out_p, n = out_n)
  }

  if(delete_upper_tria & method == "pearspear") warning(paste(
    "You selected method 'pearspear' (Pearson above and Spearman below",
    "diagonal) and set 'delete_upper_tria = TRUE'. This is probably not",
    "what you want."
  ))

  if (is.logical(bold) && bold) pvalue <- 0.05
  else if (is.numeric(bold)) {
    pvalue <- bold
    bold <- TRUE
  }

  mat <- as.matrix(data[, sapply(data, is.numeric), drop = FALSE])
  if (method != "pearspear") {
    clist <- cor_list(mat, method)
  } else {
    pcl <- cor_list(mat, "pearson")
    scl <- cor_list(mat, "spearman")
    pcl$r[lower.tri(pcl$r)] <- scl$r[lower.tri(scl$r)]
    pcl$p[lower.tri(pcl$p)] <- scl$p[lower.tri(scl$p)]
    clist <- pcl
  }

  if (n_note) {
    n_note <- sprintf(n_note_sprintf, min(clist$n), max(clist$n))
    if (is.null(notes)) notes <- n_note
    else notes <- paste(notes, n_note)
  }

  if (bold) {
    bold_str <- sprintf(bold_sprintf, pvalue)
    if (!is.null(notes)) notes <- paste(notes, bold_str)
    else notes <- bold_str
  }

  if (!is.null(var_rename)) {
    var_names <- var_rename[row.names(clist$r)]
  } else {
    var_names <- row.names(clist$r)
  }

  if (cols_alpha) {
    row_names <- paste0(LETTERS[seq_along(var_names)], ": ", var_names)
    col_names <- c(' ', LETTERS[seq_along(var_names)])
  } else {
    row_names <- var_names
    col_names <- c(' ', var_names)
  }

  if(delete_diag) diag(clist$r) <- NA

  out <- modelsummary:::correlation_clean(clist$r, fmt = fmt, triangle = FALSE)
  if(delete_upper_tria) out[upper.tri(out)] <- '.'

  out <- cbind(rowname = row_names, out)
  colnames(out) <- col_names
  align <- paste0('l', strrep('r', ncol(out) - 1))

  kab <- modelsummary:::factory(
    out, align = align, hrule = NULL, notes = notes,
    output = 'default', title = title
  )

  if (bold) {
    for(i in 1:ncol(mat))
      kab <- kab %>% column_spec(column = i + 1, bold = clist$p[, i] < pvalue)
  }
  kab
}


# Some Examples

kable_corr_table(mtcars)

mtcars_na <- as.data.frame(apply(
  mtcars, 2, function(x) {
    n <- length(x)
    x[sample(1:n, floor(n/10))] <- NA
    x
  }
))

kable_corr_table(mtcars_na, method = "spearman", n_note = TRUE)
kable_corr_table(
  mtcars, method = "pearspear", n_note = TRUE, delete_diag = TRUE
)
kable_corr_table(
  mtcars, method = "pearspear", n_note = TRUE, delete_diag = TRUE,
  bold = TRUE
)
kable_corr_table(
  mtcars, method = "pearspear", n_note = TRUE, delete_diag = TRUE,
  bold = 0.01
)
kable_corr_table(
  mtcars, method = "pearspear", n_note = TRUE, delete_diag = TRUE,
  bold = 0.01, cols_alpha = TRUE
)
kable_corr_table(
  mtcars, method = "pearspear", n_note = TRUE, delete_diag = TRUE,
  bold = 0.01, cols_alpha = TRUE,
  note = "This is a wonderful correlations table.",
  var_rename = c(
    mpg = "Miles/US Gallon",
    cyl = "Number of cylinders",
    disp = "Displacement (cu.in.)",
    hp = "Gross horsepower",
    drat = "Rear axle ratio",
    wt = "Weight (lb/1000)",
    qsec = "1/4 mile time",
    vs = "Straight Line engine?",
    am = "Manual transmission?",
    gear = "Number of forward gears",
    carb = "Number of carburetors"
  )
)
```

**Comment by vincentarelbundock** [comment]
### Bolding p.value

One difficulty with doing this in the user-space is that after we process the table, users only have access to the character-transformed versions of the correlation coefficients, and not to any numeric values. One options would be to add an attribute to the dataset, which would be a data.frame of the same shape, but with raw numeric p-values. This would make it easy for people to extract that information and bold accordingly. s

Is there any easy way to produce a matrix of p.value of the same shape and with the same positions as the matrix of correlation coefficients? Can we do this generally for all correlation methods?

Alternatively, we could think of a `stars` argument, similar to the one in `modelsummary`.

### Renaming

For renaming columns, I think the easiest thing is to rename the columns beforehand using backticks. Maybe there should be something in the vignette and docs about this:

``` r
library("modelsummary")
library("tidyverse")

dat <- mtcars %>%
  select(`Miles / gallon` = mpg,
         `Displacement` = disp)

datasummary_correlation(dat, output = "markdown")
```

<table>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Miles / gallon</th>
<th style="text-align: right;">Displacement</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Miles / gallon</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Displacement</td>
<td style="text-align: right;">-.85</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>

### Alpha columns

Similarly for the alpha columns, a simple implementation could be like this:

``` r
# alpha columns
note <- sprintf("(%s) %s", letters[1:ncol(dat)], colnames(dat))
note <- paste(note, collapse = "; ")
colnames(dat) <- sprintf("(%s)", letters[1:ncol(dat)])

datasummary_correlation(dat, output = "markdown", notes = note)
```

<table>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">(a)</th>
<th style="text-align: right;">(b)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(a)</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">.</td>
</tr>
<tr class="even">
<td style="text-align: left;">(b)</td>
<td style="text-align: right;">-.85</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>

**Note:** (a) Miles / gallon; (b) Displacement

I can see how this can be useful, but I'm not 100% convinced yet that this use-case is broad enough to justify adding an argument: It's only 3 lines of code that users can easily wrap in a function for reproducibility.

To repeat, I'm totally open to discussion on this, but limiting the number of arguments is pretty important, since I don't want to head the `xtable` route where functions have 100+ arguments. Often, a flexible API and good documentation seem better. (But that's my bias, and I'm definitely open to suggestions.)

**Comment by vincentarelbundock** [comment]
I added some info and examples to the `?datasummary_correlation` documentation:

```r
datasummary_correlation      package:modelsummary      R Documentation

Generate a correlation table for all numeric variables in your dataset.

Description:

     The names of the variables displayed in the correlation table are
     the names of the columns in the `data`. You can rename those
     columns (with or without spaces) to produce a table of
     human-readable variables.

Examples:

     library(modelsummary)

     # clean variable names (base R)
     dat <- mtcars[, c("mpg", "hp")]
     colnames(dat) <- c("Miles / Gallon", "Horse Power")
     datasummary_correlation(dat)

     # clean variable names (tidyverse)
     library(tidyverse)
     dat <- mtcars %>%
       select(`Miles / Gallon` = mpg,
              `Horse Power` = hp)
     datasummary_correlation(dat)

     # alternative methods
     datasummary_correlation(dat, method = "pearspear")

     # custom function
     cor_fun <- function(x) cor(x, method = "kendall")
     datasummary_correlation(dat, method = cor_fun)

     # rename columns alphabetically and include a footnote for reference
     note <- sprintf("(%s) %s", letters[1:ncol(dat)], colnames(dat))
     note <- paste(note, collapse = "; ")

     colnames(dat) <- sprintf("(%s)", letters[1:ncol(dat)])

     datasummary_correlation(dat, notes = note)
```

**Comment by joachim-gassen** [comment]
I agree. Your select/rename option for customizing the variable names and col/row heads clearly suffices. I very much like your approach to keeping the API clean and tidy!

About the bolding: Yes. The fact that whatever `datasummary_correlation()` returns is character complicates conditional formatting.

> Is there any easy way to produce a matrix of p.values of the same shape and with the same positions as the matrix of correlation coefficients? Can we do this generally for all correlation methods?

In the code above `cor_list)` returns a list of rho/n/pvalue matrices. The `rho` matrix can be provided to `datasummary_correlation()` (wrapped in a function) via the `method` argument and the `pvalue` matrix can then be used for post-processing the return object from  `datasummary_correlation()`. So, we can create the table without any change to the package but this requires us to calculate the  the correlations in user-space. It would be neat if this could be done within `datasummary_correlation()` directly but to me this looks as if this would require implementing bolding for the different factory implementations, correct?

**Comment by vincentarelbundock** [comment]
OK, so here's a fun example. I use `psych::corr.test` to calculate p.values and `kableExtra::cell_spec` to change the background of cells that are significant at p<0.05. 

I think this is still too complicated for the "average" user, but it suggests a nice simple path forward.

Note that this will only work with the very latest version of `modelsummary` (from Github as of 27 seconds ago), as we need to pass `escape=TRUE` to `factory_kableExtra`, and that wasn't possible before.

```r
# use kableExtra and psych to color significant cells
library(modelsummary)
library(psych)
library(kableExtra)

dat <- mtcars[, c("vs", "hp", "gear")]

cor_fun <- function(dat) {
 # compute correlations and format them
 correlations <- data.frame(cor(dat))
 correlations <- datasummary_correlation_format(correlations, fmt = 2)

 # calculate pvalues using the `psych` package
 pvalues <- psych::corr.test(dat)$p

 # use `kableExtra::cell_spec` to color significant cells
 for (i in 1:nrow(correlations)) {
   for (j in 1:ncol(correlations)) {
     if (pvalues[i, j] < 0.05 && i != j) {
       correlations[i, j] <- cell_spec(correlations[i, j], background = "pink")
     }
   }
 }
 return(correlations)
}

# The `escape=FALSE` is important here!
datasummary_correlation(dat, method = cor_fun, escape = FALSE)
```

<img width="192" alt="Screen Shot 2021-03-14 at 09 45 18" src="https://user-images.githubusercontent.com/987057/111070864-08178d00-84aa-11eb-92f0-06f42a8a2580.png">

**Comment by joachim-gassen** [comment]
Very cool! One small thing that I would like to suggest: Currently, the code of `datasummary_correlation()` imposes column heads and row heads in the correlation table to be identical. This could be made more flexible by changing lines 149-151 from
```
  out <- cbind(rowname = row.names(out), out)

  colnames(out) <- c(' ', out[[1]])
```

to

```
col_names <- col.names(out)  
out <- cbind(rowname = row.names(out), out)
colnames(out) <- c(' ', col_names)
```

**Comment by vincentarelbundock** [comment]
I'm not sure I understand the difference. What's the use-case?

**Comment by joachim-gassen** [comment]
In most correlation tables that use the abbreviated column heads, the row heads provide the full variable name and not just (a), (b) like in your example above. This is currently not possible because column and row names are both set to the row names of the matrix. When you change the code then the method function can return a matrix with different row and column heads that are then reflected in the table.

**Comment by joachim-gassen** [comment]
Something like this:

|  | (a) | (b)|
| --------------------- |----:|-----:|
| (a) Miles/US Gallon       |   | -.85 | 
| (b) Number of cylinders | -.91 |  |

**Comment by vincentarelbundock** [comment]
Ha! I understand now!

The latest version should work with different row.names and colnames.

Now that I've seen a few, I kind of like those alpha numeric tables. I wonder if I shouldn't make this the default for any table where the total number of characters in column names exceeds ~120 char.

I'm OK with the package being opinionated, and people would always have the option of using a custom method function anyway.

**Comment by joachim-gassen** [comment]
Great!

About making alpha coded table the default when cumulated variable char length exceeds some threshold: Your call. I certainly see the advantage that correlation tables would look nicer by default, but, personally, I am not a big fan of 'smart' functions that change their default behavior based on some data condition. Think about a setting where tables are automatically generated and included in a paper. The author has fiddled around with the table's latex implementation a little bit to make sure that it renders exactly as they like it. Now they change one variable name to be one character longer and boom, the whole design of the table changes (maybe even without them noticing as the code runs through without feedback - this could be addressed by a message). I would vote for a function parameter (`col_index`?) that could be NULL or contain a simple formatting string like "(1)" "(A)" "(a)" "A:" "a:" "1:". But maybe this is over the top. So as I said: Your call  😉

**Comment by vincentarelbundock** [comment]
Glad I brainstormed this, because your argument fully convincing and I now see that it would have been a really bad idea to implement this automagical function.

You will see that I closed the `notes` is a function issue for reasons described therein.

Aside from that, I think we have enough general mechanisms to accomodate most users without too much pain.

The alpha column feature is super neat, but it feels a bit niche, so I doubt I will include a new argument just for that. However, since we can achieve the same result with `method` argument already, it might be nice to keep around a convenience function to do this.

In this file, I package some unexported functions that do these kinds of "semi-official" but undocumented things (with little promise of support). If we kept one there, users could something like:

```r
datasummary_correlation(mtcars,
  method = modelsummary:::correlation_alpha)
```
